src/storage/src/hummock/compactor/fast_compactor_runner.rs

```diff
@@ -0,0 +1,570 @@
+        };
+        let sst_builder = CapacitySplitTableBuilder::new(
+            builder_factory,
+            context.compactor_metrics.clone(),
+            Some(task_progress.clone()),
+            task_config.is_target_l0_or_lbase,
+            task_config.split_by_table,
+            task_config.split_weight_by_vnode,
+        );
+        assert_eq!(task.input_ssts.len(), 2);
+        let left = Box::new(ConcatSstableIterator::new(
+            task.input_ssts[0].table_infos.clone(),
+            context.sstable_store.clone(),
+            task_progress.clone(),
+        ));
+        let right = Box::new(ConcatSstableIterator::new(
+            task.input_ssts[1].table_infos.clone(),
+            context.sstable_store,
+            task_progress.clone(),
+        ));
+
+        Self {
+            executor: CompactTaskExecutor::new(sst_builder, task_config),
+            left,
+            right,
+            task_id: task.task_id,
+            metrics: context.compactor_metrics.clone(),
+            compression_algorithm,
+            task_progress,
+        }
+    }
+
+    pub async fn run(mut self) -> HummockResult<Vec<LocalSstableInfo>> {
+        self.left.rewind().await?;
+        self.right.rewind().await?;
+        let mut skip_raw_block_count = 0;
+        let mut skip_raw_block_size = 0;
+        while self.left.is_valid() && self.right.is_valid() {
+            let ret = self
+                .left
+                .current_sstable()
+                .key()
+                .cmp(&self.right.current_sstable().key());
+            let (first, second) = if ret == Ordering::Less {
+                (&mut self.left, &mut self.right)
+            } else {
+                (&mut self.right, &mut self.left)
+            };
+            assert!(ret != Ordering::Equal);
+            if first.current_sstable().iter.is_none() {
+                let right_key = second.current_sstable().key();
+                while first.current_sstable().is_valid() {
+                    let full_key = FullKey::decode(first.current_sstable().next_block_largest());
+                    // the full key may be either Excluded key or Included key, so we do not allow
+                    // they equals.
+                    if full_key.user_key.ge(&right_key.user_key) {
+                        break;
+                    }
+                    let smallest_key =
+                        FullKey::decode(first.current_sstable().next_block_smallest());
+                    if self
+                        .executor
+                        .last_key
+                        .user_key
+                        .as_ref()
+                        .eq(&smallest_key.user_key)
+                    {
+                        // If the last key is delete tombstone, we can not append the origin block
+                        // because it would cause a deleted key could be see by user again.
+                        break;
+                    }
```

Comment:
IIUC, we will `add_raw_block` if the block is in between `executor.last_key` and `right_key`. Because `add_raw_block` will close the pending block before adding the raw block, will this generate small blocks. For example, will it generate a block containing only a single key `executor.last_key`?

Reply:
Emmm, but i think the most block is still  large block. A few small block does not affect too much.
Anyway, we can check the size of current block and make sure that it is not too small.

Reply:
But it means that we need to change the capacity of block builder  and I think it is hard to write....

Reply:
A simple way is not compress this small block and it could save some resource....